File: csharp/Carballo/Com/Alonsoruibal/Chess/Movegen/LegalMoveGenerator.cs
/*
 * Code converted with Sharpen
 * 
 */using Com.Alonsoruibal.Chess;
using Com.Alonsoruibal.Chess.Movegen;
using Sharpen;

namespace Com.Alonsoruibal.Chess.Movegen
{
	public class LegalMoveGenerator : MagicMoveGenerator
	{
		/// <summary>get only LEGAL Operations by testing with domove</summary>
		public override int GenerateMoves(Board board, int[] moves, int index)
		{
			int lastIndex = base.GenerateMoves(board, moves, index);
			int j = index;
			for (int i = 0; i < lastIndex; i++)
			{
				if (board.DoMove(moves[i], false))
				{
					moves[j++] = moves[i];
					board.UndoMove();
				}
			}
			return j;
		}
	}
}

File: core/src/main/java/com/alonsoruibal/chess/movegen/LegalMoveGenerator.java
package com.alonsoruibal.chess.movegen;

import com.alonsoruibal.chess.Board;
import com.alonsoruibal.chess.Move;

public class LegalMoveGenerator extends MagicMoveGenerator {

	/**
	 * Get only LEGAL moves testing with doMove
	 * The moves are returned with the check flag set
	 */
	@Override
	public int generateMoves(Board board, int[] moves, int index) {
		int lastIndex = super.generateMoves(board, moves, index);
		int j = index;
		// Iterate over the freshly generated pseudo-legal moves [index, lastIndex),
		// not from 0: the slots below `index` contain stale data from previous fills
		// and re-running doMove on them would corrupt board state and emit bogus moves.
		for (int i = index; i < lastIndex; i++) {
			if (board.doMove(moves[i], true, false)) {
				moves[j++] = board.getCheck() ? moves[i] | Move.CHECK_MASK : moves[i];
				board.undoMove();
			}
		}
		return j;
	}
}